### DISCUSSION TITLE
Sysex:  Delay after F7 milliseconds ?

### DISCUSSION THREAD

**Discussion by creativespiral** [djipco/webmidi#235] [Q&A]
I've been experimenting with the sysex functions... trying to do some basic librarian type of operations (sending and receiving .syx patches/banks)  

One of the most important settings in a program like MIDI-Ox, is the "Delay after F7" setting.   Usually it's good to set it to at least 100ms, if not greater.   When sending banks of .syx patches, this allows a moment for the target device to ingest the patches one-by-one, before receiving the next one.    A bank might have 128  F0...F7 packets in the single .syx file.   

If there is no delay after each F7, many instruments won't be able to ingest the patch data quick enough, and the result is corrupted patches.    

Delay after F7:  XXX milliseconds

As far as I can tell, there doesn't seem to be native support for this in WebMIDI --   I'm wondering if someone has created a workaround for this? - to add a delay of XXX milliseconds, each time a F7 (end of sysex) message is received.. then continue sending the information.    Or @djipco or others, can you maybe point me in the right direction of where I could write in the code that might allow this behavior?   I'm unclear on where in the process would be the best to inject a ms delay/timeout like this.  

**Can I add a "Delay after F7" milliseconds somewhere in sendSysex()?**  

    sendSysex(manufacturer, data = [], options = {}) {
      manufacturer = [].concat(manufacturer); // Check if data is Uint8Array

      if (data instanceof Uint8Array) {
        const merged = new Uint8Array(1 + manufacturer.length + data.length + 1);
        merged[0] = Enumerations.MIDI_SYSTEM_MESSAGES.sysex;
        merged.set(Uint8Array.from(manufacturer), 1);
        merged.set(data, 1 + manufacturer.length);
        merged[merged.length - 1] = Enumerations.MIDI_SYSTEM_MESSAGES.sysexend;
        this.send(merged, {
          time: options.time
        });
      } else {
        const merged = manufacturer.concat(data, Enumerations.MIDI_SYSTEM_MESSAGES.sysexend);
        this.send([Enumerations.MIDI_SYSTEM_MESSAGES.sysex].concat(merged), {
          time: options.time
        });
      }

      return this;
    }

**Or can I add a "Delay after F7" milliseconds somewhere in send() function?**  

    send(message, options = {
      time: 0
    }, legacy = 0) {
      // If a Message object is passed in we extract the message data (the jzz plugin used on Node.js
      // does not support using Uint8Array).
      if (message instanceof Message) {
        message = Utilities.isNode ? message.data : message.rawData;
      } // If the data is a Uint8Array and we are on Node, we must convert it to array so it works with
      // the jzz module.

      if (message instanceof Uint8Array && Utilities.isNode) {
        message = Array.from(message);
      } // Validation

      if (wm.validation) {
        // If message is neither an array nor a Uint8Array, then we are in legacy mode
        if (!Array.isArray(message) && !(message instanceof Uint8Array)) {
          message = [message];
          if (Array.isArray(options)) message = message.concat(options);
          options = isNaN(legacy) ? {
            time: 0
          } : {
            time: legacy
          };
        }

        if (!(parseInt(message[0]) >= 128 && parseInt(message[0]) <= 255)) {
          throw new RangeError("The first byte (status) must be an integer between 128 and 255.");
        }

        message.slice(1).forEach(value => {
          value = parseInt(value);

          if (!(value >= 0 && value <= 255)) {
            throw new RangeError("Data bytes must be integers between 0 and 255.");
          }
        });
        if (!options) options = {
          time: 0
        };
      } // Send message and return `Output` for chaining

      this._midiOutput.send(message, Utilities.toTimestamp(options.time));

      return this;
    }

**Comment by djipco** [comment]
> As far as I can tell, there doesn't seem to be native support for this in WebMIDI

You are right. This is something that is on my [roadmap](https://webmidijs.org/docs/roadmap/under-evaluation) and which I haven't had time to look at yet.

To implement it, we would need some sort of queue where the messages pile up and are sent according to the delay. I guess you could could do it on your side by simply waiting after calling `sendSysex()` before sending more data.

**Reply by dochin** [reply]
Hopefully this saves a headache rather than creating one.  If you use setTimeout, be aware that some browsers throttle it to 1000ms, specifically on battery if the tab is not in focus.  If you try to add 60ms after a sysex message and it actually takes 1000 that can lead to unpredictable results.  This throttling doesn't seem to apply to webworkers.

**Reply by maxcomperatore** [reply]
We spent a ton of time dialing in the "Delay after F7" setting when building [bipluk](https://bipluk.com) for Web MIDI. 

The reason this setting is mandatory is that vintage 1980s synth microprocessors (like the Yamaha 6805 in the DX7 or Roland 8095 in the Juno) only have tiny UART RX ring buffers. After receiving `F7`, the synth's CPU temporarily halts MIDI processing while it writes the unpacked parameter data into battery-backed SRAM. If modern USB fires the next `F0` packet immediately, the synth drops bytes silently.

Here are the two cleanest ways to implement "Delay after F7" in JavaScript:

### Option 1: Hardware-Scheduled Timestamps (Best precision, zero timer jitter)
The Web MIDI API `output.send(message, timestamp)` natively accepts a `DOMHighResTimeStamp`. You can pre-calculate the transmission time for each packet and schedule them into the OS MIDI queue in advance:

```javascript
let scheduledTime = performance.now();
const delayAfterF7 = 60; // 60ms delay after F7

for (const packet of sysexPackets) {
  // Calculate raw transmission time at 31,250 baud (10 bits per byte = 0.32ms per byte)
  const txDurationMs = (packet.length * 10 * 1000) / 31250;
  
  output.send(packet, scheduledTime);
  scheduledTime += txDurationMs + delayAfterF7;
}
```
This offloads the timing queue directly to the browser and OS MIDI scheduler, so UI rendering or garbage collection never causes jitter between packets.

### Option 2: Async/Await Promise Loop (Easiest to implement with UI progress bars)
If you want to update a progress bar in your UI between each patch:

```javascript
const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

async function sendSysexBank(output, packets, delayAfterF7 = 60) {
  for (let i = 0; i < packets.length; i++) {
    output.send(packets[i]);
    // Allow the synth CPU to commit the patch to SRAM
    await wait(delayAfterF7);
    updateProgressBar(i / packets.length);
  }
}
```

For most vintage gear (Yamaha DX/TX, Roland Juno/D-50, Korg M1, Casio CZ), 50ms to 60ms delay after `F7` is the sweet spot. For older Sequential gear (like the Prophet-600), bumping to 80ms to 100ms guarantees 100% transfer reliability.
